add some more time idioms

maxbachmann opened the pull request:
## Description
- add some time idioms like 'viertel vier'/ 'quarter past three' for the nice_time function, which was a TODO
- improve code readability

## How to test
The changes to the german version are covered by unit tests.
For the english version there are no unit tests yet.

## Contributor license agreement signed?
CLA [x] signed now

devs-mycroft commented:
Hello, @maxbachmann, thank you for helping with the Mycroft project! We welcome everyone
into the community and greatly appreciate your help as we work to build an AI
for Everyone.

To protect yourself, the project, and users of Mycroft technologies we require
a Contributor Licensing Agreement (CLA) before accepting any code
contribution. This agreement makes it crystal clear that along with your
code you are offering a license to use it within the confines of this project.
You retain ownership of the code, this is just a license.

Please visit https://mycroft.ai/cla to initiate this one-time signing. Thank
you!

forslund commented:
Thanks @maxbachmann It does look good. Could you add a couple of testcases for the english half past and quarter past (in `test/unittests/util/test_format.py`)

Maybe @krisgesling or @chrisveilleux who are native english speakers can do the final review on the changes so they're all ok. (I'm unsure about using 12 for midnight for example but I'm used to 24 hour clocks)

maxbachmann commented:
@forslund sure I can add testcases for the english part later.

on the 12 hour times I was unsure aswell.
I guess in german I would usually say:
  halb/dreiviertel/halb/viertel nach 12
but
 0 Uhr zwanzig (or even 10 vor halb eins)

maxbachmann commented:
For comparision: https://github.com/project-alice-powered-by-snips/ProjectAliceModules/blob/master/PublishedModules/Psychokiller1888/DateDayTimeYear/DateDayTimeYear.py is how I currently implemented this for ProjectAlice (without am/pm support right now)

krisgesling commented:
Haven't reviewed yet, but 12 is a normal way of talking about midnight including:
> 5 to 12
> half past 12
> 12 midnight
> 12 o'clock tonight
> quarter to 12

forslund commented:
Thanks for verifying @krisgesling, always good to have the input of someone who speaks the language everyday. I don't quite trust my own skills with the language.

@maxbachman I'll wait for the extra test cases, after that we can merge this.

Many thanks for contributing!

maxbachmann commented:
I added unittests however I am not sure on two things:

maxbachmann commented on test/unittests/util/test_format_en.py:
```diff
@@ -0,0 +1,370 @@
+                         "zero zero zero two")
+        self.assertEqual(nice_time(dt, use_24hour=True,
+                                   use_ampm=False),
+                         "zero zero zero two")
```
is the time really spoken this way? with the zero sounds awkward to me, but I am not a antive english speaker

maxbachmann replied:
Yup apparently it is google tells me xD

maxbachmann commented on test/unittests/util/test_format_en.py:
```diff
@@ -0,0 +1,370 @@
+        self.assertEqual(pronounce_number(-21.234, places=4),
+                         "minus twenty one point two three four")
+        self.assertEqual(pronounce_number(-21.234, places=5),
+                         "minus twenty one point two three four")
```
why does it add `null` in german  when the precision is higher than the precision of the given number, while in english it does not add `zero` and `zero zero` in the end?

ChanceNCounter replied:
"minus twenty one" is a perfectly ordinary, colloquial way of phrasing it in English, but I should point out that (at least to an American) it comes across as extremely informal, as colloquial as it gets. It might be worth considering, "negative twenty one," which is the most common phrasing.

"Minus x," in my experience, is most frequently used when referring to weather. Two minus three is negative one, but a cold temperature would be, "minus one degree."

maxbachmann replied:
thats not what I mean:
I mean why in german it is:
`minus einundzwanzig komma zwei drei vier null null`
and in english
`minus twenty one point two three four`
and not 
`minus twenty one point two three four zero zero`
when precision is set to 5

maxbachmann commented:
since I already cleaned up the unicode strings and file encoding for the files I updated, I did this globally (should this be in the same PR?)
Another thing I wondered is why does /doc/conf.py enforce python2? It thought the project dropped python2 support

forslund commented:
It'd be better if it was in a separate PR but this is fine.

doc/conf.py is not used as a main file so it doesn't really matter but it seems like it was a change we missed when upgrading to python 3.

Can you join your English tests with the rest of the english tests in `test_format.py`?

maxbachmann commented:
@forslund oh I did not realise you already had english test cases. I added them to the correct file.

The python2 cleanup changes can be found in the https://github.com/MycroftAI/mycroft-core/pull/2396 now

Is the beahviour I described in my code comment correct? It appears to me, that it is wrong in one of the two languages

maxbachmann commented:
should the same PR be created for lingua_franca?

forslund commented:
Thank you.

Hmm, the number of places thing? from thinking about it briefly the english looks wrong...But reading the docstring in the format_en.py file says: 

`places(int): maximum decimal places to speak`

which makes the english version seem to follow the specification. On the other hand the entry function in `format.py` doesn't document that argument so technically anything goes but it should be specified to something...

We could raise a separate issue for that.

I think this is a decent state now. I'll do a final look over tomorrow morning and merge it.

maxbachmann commented:
I added an issue on the topic: https://github.com/MycroftAI/mycroft-core/issues/2397

forslund commented:
Changes looks good! Thanks for rebasing and providing a nice clean update! Merging

forslund commented:
And yes, I think you should create the same PR for lingua franca, the intention is to switch over to using that instead of the internal utils in the future.
